### PR TITLE
gem5 integration in lisa

### PR DESCRIPTION
**Pull request by avanlaer**
gem5: Added an if-statement in board configuration, to allow for the setup of a gem5 platform

This needs to be combined with changes in devlib (right now, commit 9b41c47 in github.com/avanlaer/devlib but more commits might follows).
The extra code needed in lisa might change, if the underlying code in devlib changes but this might give you an idea.

 Changes to be committed:
	modified:   libs/utils/env.py

### CODE REVIEW

File: libs/utils/env.py [hunk 1]
```diff
@@ -379,6 +379,14 @@ def _init_target(self, force = False):
             platform = Platform(model='MT8173')
             self.__modules = ['bl', 'cpufreq']
 
+        # Initialize gem5 simulation as a board
+        elif self.conf['board'].upper() == 'GEM5':
+            platform = devlib.platform.arm.Gem5SimulationPlatform(name='gem5',
```
Comment (ionela-voinescu) [inline libs/utils/env.py]: This has now changed to be `devlib.platform.gem5.Gem5SimulationPlatform`.

**Comment by avanlaer** [comment]
I've updated the commit to account for the change in devlib.

File: libs/utils/env.py [hunk 2]
```diff
@@ -379,6 +379,14 @@ def _init_target(self, force = False):
             platform = Platform(model='MT8173')
             self.__modules = ['bl', 'cpufreq']
 
+        # Initialize gem5 simulation as a board
+        elif self.conf['board'].upper() == 'GEM5':
+            platform = devlib.platform.gem5.Gem5SimulationPlatform(name='gem5',
+                       gem5_bin=self.conf['gem5_bin'],
+                       gem5_args=self.conf['gem5_args'],
+                       gem5_virtio=self.conf['gem5_virtio'],
+                       host_output_dir=self.conf['host_output_dir'],
```
Comment (derkling) [inline libs/utils/env.py]: Cannot this be set to the already defined `res_dir`?

File: libs/utils/env.py [hunk 3]
```diff
@@ -379,6 +379,14 @@ def _init_target(self, force = False):
...
+                       gem5_telnet_port=self.conf['gem5_telnet_port'] if 'gem5_telnet_port' in self.conf else None)
```
Comment (derkling) [inline libs/utils/env.py]: Better:

```python
self.conf.get('gem5_telnet_port', None)
```

File: libs/utils/env.py [hunk 4]
```diff
@@ -860,7 +868,6 @@ def install_kernel(self, tc, reboot=False):
                 self._log.warning('Using pre-installed DTB')
             else:
                 self.tftp_deploy(tc['dtb'])
-
```
Comment (derkling) [inline libs/utils/env.py]: Not part of this commit?

File: libs/utils/env.py [hunk 5]
```diff
@@ -379,6 +379,14 @@ def _init_target(self, force = False):
             platform = Platform(model='MT8173')
             self.__modules = ['bl', 'cpufreq']
 
+        # Initialize gem5 simulation as a board
+        elif self.conf['board'].upper() == 'GEM5':
+            platform = devlib.platform.gem5.Gem5SimulationPlatform(name='gem5',
+                       gem5_bin=self.conf['gem5_bin'],
```
Comment (derkling) [inline libs/utils/env.py]: What abut using a dictionary in target.conf

```python
my_conf = {
    ....
    "board" = "gem5",
    "gem5" = {
        "bin" = "...",
        "virtio" = "...",
        "args" = "...",
    }
```

And perhaps we can also "force" the things by assuming that if you have 'gem5' in conf then this is the board to use... but let see what @bjackman think about that ;-)

File: libs/utils/env.py (hunk 3 above)
Comment (bjackman) [inline libs/utils/env.py]: Or just `self.conf.get('gem5_telnet_port')` as `dict.get` has a `None` default by default

File: libs/utils/env.py (hunk 5 above)
Comment (bjackman) [inline libs/utils/env.py]: Agree that we should have a 'gem5' subdict and not 'gem5_*'.

It's unfortunate to have to have a generic "gem5" board as this board is special and violates assumptions that you can make about the other boards (e.g. if you see "pixel" you know it's an arm64 board with topology XYZ etc, but if you see "gem5" it could be anything, you need to examine the config further). 

It would be simpler to instead have a fixed set of gem5 boards like "gem5_myplatform", "gem5_myotherplatform" and require adding their parameters somewhere in the LISA source. However gem5 boards are naturally going to be a) ephemeral experimental HW designs or b) confidential future HW designs and neither of these cases plays nicely with coding simulation parameters into the LISA source. 

So let's keep it this way - I just wanted to share my thinking; we'll need to be aware that there's a new "special" type of board and try to find any assumptions in the code that this violates.

**Comment by ionela-voinescu** [comment]
Thank  you @avanlaer for this pull request. We've created another pull request at https://github.com/ARM-software/lisa/pull/382 and we'll continue reviewing gem5 support there.
